### ISSUE TITLE
Missing values in classifier

### ISSUE THREAD

**Issue by alextamburino** [cole-trapnell-lab/garnett#14]
Hi,

I recently added 'not expressed: Gene X' to my marker lists and the created classifiers containing missing values (see below). The same dataset and 'expressed: Genes' worked prior to this.

What could the cause be? Can the same not expressed gene be used multiple times?

> classifier <- train_cell_classifier(cds = cds,marker_file = "finalMarkers.txt",
						 db=org.Mm.eg.db, cds_gene_id_type = "SYMBOL",							 
                                                 num_unknown = 500, marker_file_gene_id_type = "SYMBOL")
There are 8 cell type definitions
training_sample
        B cells          B16F10 Dendritic cells       Monocytes     Neutrophils
            500             437             500             500             500
       NK cells   Stromal cells         T cells         Unknown
             93              18             500             500
Model training finished.
Warning messages:
1: Removed 8 rows containing missing values (geom_point).
2: Removed 8 rows containing missing values (geom_text_repel).

> sessionInfo()
R version 3.5.2 (2018-12-20)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Red Hat Enterprise Linux

Matrix products: default
BLAS/LAPACK: /SFS/product/R/3.5.2/centos76_x86_64/extlibs/lib/libopenblas_nehalemp-r0.3.5.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
 [1] splines   stats4    parallel  stats     graphics  grDevices utils
 [8] datasets  methods   base

other attached packages:
 [1] reticulate_1.11.1    dplyr_0.8.0.1        reshape_0.8.8
 [4] org.Mm.eg.db_3.7.0   AnnotationDbi_1.44.0 IRanges_2.16.0
 [7] S4Vectors_0.20.1     garnett_0.1.4        monocle_2.9.0
[10] DDRTree_0.1.5        irlba_2.3.3          VGAM_1.1-1
[13] ggplot2_3.1.1        Biobase_2.42.0       BiocGenerics_0.28.0
[16] Matrix_1.2-17        Seurat_3.0.0.9000

**Comment by alextamburino** [comment]
I went back and retested without the 'not expressed' genes and still have the issue. I just installed Monocle 3 Beta this week. Would that have altered any of the dependencies for Garnett?

**Comment by hpliner** [comment]
Are you sure those warnings are from training the classifier? The only place where geom_text_repel is called is in the plot_markers function - where it usually indicates a few NAs (usually because of a zero somewhere - not usually a problem).

**Comment by alextamburino** [comment]
Hi Hannah,

Sorry for the confusion. I created a wrapper that iterates the marker selection and plot_markers (based on our Google Groups discussion), so that makes sense. (I've tested the functions on their own as well).

Perhaps the subsequent error is more important:

> cds <- classify_cells(cds, classifier,db = db, cluster_extend = TRUE,cds_gene_id_type = "SYMBOL")
<simpleError in rowQ(imat, ncol(imat)): cannot handle missing values.>

Would this have resulted from Monocle 3 Beta driven updates?

Alex

**Comment by alextamburino** [comment]
Hi Hannah,

I've traced back my steps all the way from the beginning and the issue seems to have been cause by one or more genes in the marker list which didn't translate somewhere during the training or classification. I was able to run Garnett on the same file using a truncated gene set. All markers were in_cds=TRUE and had ensembl gene_id, but must have been lost somewhere (?). Curious if you have any thoughts.

Alex

**Comment by hpliner** [comment]
Hi Alex,

Garnett uses the db passed to the function to convert gene ids. Its possible that when you updated for Monocle 3 you also updated the db version you were using which might have introduced this? You can check whether the db has all of the genes you're looking for using the AnnotationDbi functions, i.e.

```
> library(org.Mm.eg.db)
Loading required package: AnnotationDbi

> keytypes(org.Mm.eg.db)
 [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS" "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL" 
[10] "GENENAME"     "GO"           "GOALL"        "IPI"          "MGI"          "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"        
[19] "PMID"         "PROSITE"      "REFSEQ"       "SYMBOL"       "UNIGENE"      "UNIPROT"     
> select(org.Mm.eg.db, "Myog", "ENSEMBL", "SYMBOL")
'select()' returned 1:1 mapping between keys and columns
  SYMBOL            ENSEMBL
1   Myog ENSMUSG00000026459

```

If they're all in there then we'll have to do a bit more digging, but that's the first thought that comes to mind.

Hannah

**Comment by alextamburino** [comment]
Hi Hannah,

I was able to remove two genes from the marker_file and get it work. No rationale for why they don't work. Those two genes work on the same dataset when used in combination with other genes which is even weirder.

I just ran into another instance of the similar issue where a marker_file works on one version of the dataset, but not another (just different normalizations). The error is repeatable when everything is kept the same, but seems to be sporadic otherwise.

I'll just work around it for now. Please let me know if you discover this error (or a solution) at any point.

Thanks,
Alex

**Comment by justinesjw** [comment]
Hi @alextamburino,

I'm facing the same error 
 `<simpleError in rowQ(imat, ncol(imat)): cannot handle missing values.>`
but in my case, I do not know which gene(s) is/are not expressed in my dataset. 
All the genes used are DE genes from scanpy which known are expressed. 

Can you advise me on how to identify the genes in marker file that might not be expressing? 

Thanks, 
Justine

**Comment by alextamburino** [comment]
Hi @justinesjw 

I would use marker_check to confirm AnnotationDbi recognizes and converts the gene names. Other than that I devised a scheme for systematically testing and eliminating markers. Took a long time.

Sorry I can't provide an easier solution.

Alex

**Comment by hpliner** [comment]
Hi @alextamburino,

Thanks for following up. I haven't been able to reproduce the error on my end, if you're willing to share some more information on the gene names/species that are being skipped I can try to figure out what's going on. Feel free to email it hpliner@uw.edu if you're willing.

Best,
Hannah

- Also, I'm working on adding some flags to avoid the very ambiguous rowQ error. Stay tuned

**Comment by justinesjw** [comment]
Hi @alextamburino 

Thanks for your speedy reply!

May I know your workflow of testing and eliminating markers?
Do I individually plot the expression?

Thanks,
Justine

**Comment by alextamburino** [comment]
Hi @justinesjw 
It wasn't anything fancy. I just created input marker files where I systematically deleted one gene from each cell type. I kept running this through Garnett and waited to see if I got the error.

Hi @hpliner 
I'll connect with you over email.

Thanks,
Alex

**Comment by hpliner** [comment]
@justinesjw @alextamburino 
I've maybe figured out one part of the mystery - the rowQ error reported above appears to be caused by NaNs ending up in the cds after size factor correction - in the case I was working with, this was because there were cells with zero reads. I've added a check and fix for zero read cells (monocle3 branch 0.2.0). Haven't figured out yet why NaNs would be introduced with only some markers and not others... but I'll keep looking. Feel free to test out the bug fix to see if it changes anything for your cases.
